File: python/chord_tree.py
import chord_data
from chord import Chord
import random

class PotentialChords(object):
    """tree containing all potential chords"""
    BASE_CHORDS = 'C', 'C#', 'Db', 'D', 'D#', 'Eb', 'E', 'F', 'F#', 'Gb', 'G', 'G#', 'Ab', 'A', 'A#', 'Bb', 'B'
    SPECIAL_CHORDS = 'A/C#', 'A/E', 'A/F', 'A/G', 'A/G#', 'Am/C', 'Am/E', 'Am/F', 'Am/F#', 'Am/G', 'Am/G#', 'C/E', 'C/F', 'C/G', 'D/A', 'D/B', 'D/Bb', 'D/C', 'D/F#', 'E/B', 'E/C#', 'E/D', 'E/D#', 'E/F', 'E/F#', 'E/G', 'E/G#', 'Em/B', 'Em/C#', 'Em/D', 'Em/D#', 'Em/F', 'Em/F#', 'Em/G', 'Em/G#', 'F/A', 'F/C', 'F/D', 'F/D#', 'F/E', 'F/G', 'Fm/C', 'G/B', 'G/D', 'G/E', 'G/F', 'G/F#'

    def __init__(self):
        self.head = Chord(None, None)
        self.build_tree()

    def build_tree(self):
        potential_chords = chord_data.data
        for notes in PotentialChords.BASE_CHORDS:
            self.head.add_child(Chord(notes, None))
        for child in self.head.children:
            for potential_chord in potential_chords:
                if child.chord in potential_chord['chord']:
                    Chord(potential_chord, child)

    def random_chord(self):
        chord_family = PotentialChords.BASE_CHORDS[random.randrange(0, len(PotentialChords.BASE_CHORDS))]
        potential_chords = self.chords_in_family(chord_family)
        return potential_chords[random.randrange(0, len(potential_chords))]

    def random_easy_chord(self):
        chord_family = PotentialChords.BASE_CHORDS[random.randrange(0, len(PotentialChords.BASE_CHORDS))]
        potential_chords = self.find_where({'modifier': 'major'})
        return potential_chords[random.randrange(0, len(potential_chords))]

    def chords_in_family(self, family):
        for note_family in self.head.children:
            if note_family.chord == family:
                return note_family.children

    def find_where(self, chord_options):
        matching_nodes = []
        for note_family in self.head.children:
            for chord in note_family.children:
                if chord.matches_critera(chord_options):
                    matching_nodes.append(chord)
        return matching_nodes

File: ruby/chord_tree.rb
require_relative "chord_data"
require_relative 'chord'

class PotentialChords
  attr_reader :head, :prev_song

  BASE_CHORDS = ['C', 'C#', 'Db', 'D', 'D#', 'Eb', 'E', 'F', 'F#', 'Gb', 'G', 'G#', 'Ab', 'A', 'A#', 'Bb', 'B']

  SPECIAL_CHORDS = ['A/C#', 'A/E', 'A/F', 'A/G', 'A/G#', 'Am/C', 'Am/E', 'Am/F',
                    'Am/F#', 'Am/G', 'Am/G#', 'C/E', 'C/F', 'C/G', 'D/A', 'D/B',
                    'D/Bb', 'D/C', 'D/F#', 'E/B', 'E/C#', 'E/D', 'E/D#', 'E/F',
                    'E/F#', 'E/G', 'E/G#', 'Em/B', 'Em/C#', 'Em/D', 'Em/D#',
                    'Em/F', 'Em/F#', 'Em/G', 'Em/G#', 'F/A', 'F/C', 'F/D', 'F/D#',
                    'F/E', 'F/G', 'Fm/C', 'G/B', 'G/D', 'G/E', 'G/F', 'G/F#']

  def initialize()
    @head = Chord.new(nil, nil)
    build_tree
  end

  def build_tree
    potential_chords = ChordData.data
    BASE_CHORDS.each do |chord|
      @head.add_child(Chord.new(chord, nil))
    end
    @head.children.each do |child|
      potential_chords.each do |potential_chord|
        next unless potential_chord['chord'].match(child.chord)
        Chord.new(potential_chord, child)
      end
    end
  end

  def random_chord
    chord_family = BASE_CHORDS[rand(BASE_CHORDS.length)]
    potential_chords = chords_in_family(chord_family)
    potential_chords[rand(potential_chords.length)]
  end

  def random_easy_chord
    chord_family = BASE_CHORDS[rand(BASE_CHORDS.length)]
    potential_chords = find_where(modifier: 'major')
    potential_chords[rand(potential_chords.length)]
  end

  protected
  def chords_in_family(family)
    @head.children.each do |note_family|
      next unless note_family.chord == family
      return note_family.children
    end
    nil
  end

  def find_where(chord_options)
    matching_nodes = []
    @head.children.each do |note_family|
      note_family.children.each do |chord|
        if (chord.matches_critera(chord_options))
          matching_nodes << chord
        end
      end
    end

    matching_nodes
  end

end

# base_chords: ['C', 'C#', 'Db', 'D', 'D#', 'Eb', 'E', 'F', 'F#', 'Gb', 'G', 'G#', 'Ab', 'A ', 'A#', 'Bb', B]
# special_chords: ['A/C#', 'A/E', 'A/F', 'A/G', 'A/G#', 'Am/C', 'Am/E', 'Am/F', 'Am/F#', 'Am/G', 'Am/G#', 'C/E', 'C/F', 'C/G', 'D/A', 'D/B', 'D/Bb', 'D/C', 'D/F#', 'E/B', 'E/C#', 'E/D', 'E/D#', 'E/F', 'E/F#', 'E/G', 'E/G#', 'Em/B', 'Em/C#', 'Em/D', 'Em/D#', 'Em/F', 'Em/F#', 'Em/G', 'Em/G#', 'F/A', 'F/C', 'F/D', 'F/D#', 'F/E', 'F/G', 'Fm/C', 'G/B', 'G/D', 'G/E', 'G/F', 'G/F#']


# conditions for a song to be over?
# song ends on the same chord as it starts
# song must be at least 35 chords long

# conditions for a valid chord
# chord must be played no more than 4 times in a row

# what constitues an easy path
# number of chords cannot exceed 4
# chords cannot have modifiers other than minor and major
# no special chords can be included

# what constitues a medium path
# number of chords cannot exceed 7
# between 1 and 3 chords should have modifiers other than minor and major
# songs contain one special chord

# what constitues a hard path
# number of chords must exceed 4
# at least 3 chords must have modifiers other than minor and major
# songs contain 2 special chords
